### ISSUE TITLE
Allow excluding checkers from ament_lint_auto_find_test_dependencies

### ISSUE THREAD

**Issue by jpsamper2009** [ament/ament_lint#132]
## Motivation
- As mentioned in https://github.com/ros2/rosbag2/pull/86#issuecomment-461099130, it is possible to exclude specific checkers by setting `<package>_FOUND` to `TRUE`: e.g.

```cmake
find_package(ament_lint_auto REQUIRED)
# Disable auto copyright
set(ament_cmake_copyright_FOUND TRUE)
# Disable auto cpplint
set(ament_cmake_cpplint_FOUND TRUE)
ament_lint_auto_find_test_dependencies()
```

- However, if you still want to run the checkers (with a different configuration or on a subset of files, for example), the cmake code above becomes:

```cmake
find_package(ament_lint_auto REQUIRED)
# Disable auto copyright
set(ament_cmake_copyright_FOUND TRUE)
# Disable auto cpplint
set(ament_cmake_cpplint_FOUND TRUE)
ament_lint_auto_find_test_dependencies()

# Reenable copyright
set(ament_cmake_copyright_FOUND FALSE)
find_package(ament_cmake_copyright)
ament_copyright(${ONLY_SOME_FILES})

# Reenable cpplint
set(ament_cmake_cpplint_FOUND FALSE)
find_package(ament_cmake_cpplint_FOUND)
ament_cpplint(FILTERS "-readability/inheritance")
```

- Adding and `EXCLUDE` option in `ament_lint_auto_find_test_dependencies` could make the code above cleaner: e.g.

```cmake
find_package(ament_lint_auto REQUIRED)
ament_lint_auto_find_test_dependencies(
  EXCLUDE
  ament_cmake_copyright
  ament_cmake_cpplint
)

find_package(ament_cmake_copyright)
ament_copyright(${ONLY_SOME_FILES})

find_package(ament_cmake_cpplint)
ament_cpplint(FILTERS "-readability/inheritance")
```

- My thought is that we can basically move the `set()` commands into the [macro](https://github.com/ament/ament_lint/blob/master/ament_lint_auto/cmake/ament_lint_auto_find_test_dependencies.cmake)

## Goals

- Simplify the `CMakeLists.txt` file of packages which have a checkers or files diverging from the default checkers
- Make sure that new files are automatically checked, such that they need to be exclude explicitly
- Make sure that new linters (e.g. `xmllint`) automatically get added

## Relate issues

- This feature request is similar to #119, with the difference that the goal here is not to turn off the checker, but to call the checker in a non-default way.

## Open questions

- If this feature were implemented, would it be accepted? Or is it adding unnecessary complexity to `ament_lint_auto_find_test_dependencies`?

**Comment by dirk-thomas** [comment]
I don't think the proposed approach:

> Adding and `EXCLUDE` option in `ament_lint_auto_find_test_dependencies` could make the code above cleaner: e.g.

would work. The `ament_lint_auto` extension point is being triggered when `ament_package()` is called and if that call happens after the above snippet it would still trigger all linters previous being found (independently from how they are found - by `ament_lint_auto_find_test_dependencies` or an explicit call to `find_package`).

> If this feature were implemented, would it be accepted?

Sure.

**Comment by jpsamper2009** [comment]
@dirk-thomas Thanks for the quick response!

Re:
> The ament_lint_auto extension point is being triggered when ament_package() is called and if that call happens after the above snippet it would still trigger all linters previous being found (independently from how they are found - by ament_lint_auto_find_test_dependencies or an explicit call to find_package).

So we would need to find a way to determine if a package was found by `ament_lint_auto_find_test_dependencies` or by an explicit call, correct? 

What if:

We set a variable in `ament_lint_auto_find_test_dependencies` called `REGISTER_AUTO_LINT_EXTENSIONS`, which would then be used in each linter package (ament_cmake_cpplint, etc) to conditionally register the auto lint extension: e.g.,

```cmake
macro(ament_lint_auto_find_test_dependencies)
...
  set(REGISTER_AMENT_LINT_AUTO_EXTENSIONS TRUE)
  # try to find_package() all test dependencies
  foreach(_dep ${${PROJECT_NAME}_TEST_DEPENDS})
    find_package(${_dep} QUIET)
    if(${_dep}_FOUND)
      list(APPEND ${PROJECT_NAME}_FOUND_TEST_DEPENDS ${_dep})
    endif()
  endforeach()
  set(REGISTER_AMENT_LINT_AUTO_EXTENSIONS FALSE)
endmacro()
```
Then in, `ament_cmake_cpplint`:
```cmake
if(REGISTER_AMENT_LINT_AUTO_EXTENSIONS)
  ament_register_extension("ament_lint_auto" "ament_cmake_cpplint"
    "ament_cmake_cpplint_lint_hook.cmake")
endif()
```

If a package is missing the `if()`, it will still get picked up by `ament_lint_auto_find_test_dependencies`, so we don't run the risk of accidentally turning off a checker.

Thoughts?

**Comment by dirk-thomas** [comment]
Please see my comments on the PR: #133.

**Comment by dirk-thomas** [comment]
Implemented in #133.
